apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/sub-task-agent.ts

```diff
@@ -0,0 +1,725 @@
+      llmResponse,
+    )}`,
+  );
+};
+
+const generateProgressReport = (params: {
+  input: SubTaskAgentInput;
+  state: SubTaskAgentState;
+}): LlmMessageTextContent => {
+  const { state } = params;
+
+  return {
+    type: "text",
+    text: dedent(`
+      Here is a summary of the progress you've made so far.
+
+      ${generatePreviouslyInferredFactsSystemPromptMessage(state)}
+
+      You have previously proposed the following plan:
+      ${state.plan}
+
+      If you want to deviate from this plan or improve it, update it using the "updatePlan" tool.
+      You must call the "updatePlan" tool alongside other tool calls to progress towards completing the task.
+    `),
+  };
+};
+
+const getNextToolCalls = async (params: {
+  input: SubTaskAgentInput;
+  state: SubTaskAgentState;
+}): Promise<{
+  toolCalls: ParsedLlmToolCall<SubTaskAgentToolName>[];
+}> => {
+  const { input, state } = params;
+
+  const systemPrompt = dedent(`
+      ${generateSystemPromptPrefix({ input })}
+      
+      Make as many tool calls as are required to progress towards completing the task.
+    `);
+
+  const llmMessagesFromPreviousToolCalls = mapPreviousCallsToLlmMessages({
+    previousCalls: state.previousCalls,
+  });
+
+  const lastUserMessage = llmMessagesFromPreviousToolCalls.slice(-1)[0];
+
+  if (lastUserMessage && lastUserMessage.role !== "user") {
+    throw new Error(
+      `Expected last message to be a user message, but it was: ${stringify(
+        lastUserMessage,
+      )}`,
+    );
+  }
+
+  const progressReport = generateProgressReport({ input, state });
+
+  const messages: LlmMessage[] = [
+    generateInitialUserMessage({ input }),
+    ...llmMessagesFromPreviousToolCalls.slice(0, -1),
+    lastUserMessage
+      ? ({
+          ...lastUserMessage,
+          content: [
+            ...lastUserMessage.content,
+            // Add the progress report to the most recent user message.
+            progressReport,
+          ],
+        } satisfies LlmUserMessage)
+      : [],
+  ].flat();
+
+  const tools = Object.values(
+    generateToolDefinitions({
+      omitTools: [
+        ...(state.inferredFacts.length > 0 ? [] : ["complete" as const]),
+      ],
+    }),
+  );
+
+  const { userAuthentication, flowEntityId, webId } = await getFlowContext();
+
+  const llmResponse = await getLlmResponse(
+    {
+      systemPrompt,
+      messages,
+      model,
+      tools,
+      toolChoice: "required",
+    },
+    {
+      userAccountId: userAuthentication.actorId,
+      graphApiClient,
+      incurredInEntities: [{ entityId: flowEntityId }],
+      webId,
+    },
+  );
+
+  if (llmResponse.status !== "ok") {
+    throw new Error(
+      `Failed to get LLM response: ${JSON.stringify(llmResponse)}`,
+    );
+  }
+
+  const { message } = llmResponse;
+
+  const toolCalls = getToolCallsFromLlmAssistantMessage({ message });
+
+  return { toolCalls };
+};
+
+export const runSubTaskAgent = async (params: {
```

Comment:
Not suggesting we do so now, but how much scope is there for deduplicating code between this and the top-level agent? A lot of it looks familiar but I haven't compared in detail

Reply:
There might be opportunities for further deduplication, by re-using the same method for `createInitialPlan` and `getNextToolCalls` in both agents, yes. It would require some additional work though, as the type definition for `State` and `Input` are slightly different, and the prompts adjusted conditionally based on what is available in those objects.